### ISSUE TITLE
Question: optionals in combination with querying

### ISSUE THREAD

**Issue by mad-it** [googleapis/nodejs-firestore#1357]
Hi there,

Typescript's type system has optionals built in. Here is an example:
```ts
interface Run {
  begin: Timestamp,
  end?: Timestamp
}
```

The usage of this type would look as follows:
```ts
const run1 : Run = { begin: xxx }
const run2 : Run = { begin: xxx, end: xxx } 
```

Currently with Firestore it is not possible search for all the "runs" where `end` is **not** defined. 

The **only** alternative I was able to come up with would be to completely drop the optionals of the Typescript language and use `null` as a value type in order for fields to be queryable. For example:
```ts
interface Run {
  begin: Timestamp,
  end: Timestamp | null
}
```

This does however remove the one of the basic features of Typescript. So my question is, will it ever become possible to query for fields that are present?

**Comment by merlinnot** [comment]
Firestore, as most databases, is language-independent. The representation of data does not always match 1:1, not to mention language semantics.

As an example, `Int64` and `Float64` types in Firestore are both converted to a `Number` in JavaScript.

TypeScript has a notion of optional properties, so the compiler does know that some property _potentially exists_. Firestore has no knowledge of a potential existence of that property if it's not actually set. Something that doesn't exist can't become a part of an index, and queries are performed based on indexes which allows Firestore to perform these operations efficiently.

That's also why `null` works, because it tells the database "here's a property, the name of it is X, the value is null". That way the database learns about existence of such property and can include it in an index with a value `null`.

To sum up in one sentence, it's because of a difference between "does not exist" and "exists with a value of null".

**Comment by schmidt-sebastian** [comment]
Thanks for the detailed response @merlinnot. Firestore indeed drives all queries by indices and only returns documents that match an index (and `null` values are indexed). This allows us to scale the performance of all queries with the size of the result set. 

If we instead had to scan for explicitly missing values, we would have to process all documents in a collection. As such, this is likely something we will not be able to offer anytime soon.
